File: frontend/src/api/ProductService.js
import axios from 'axios';

// Base URL for the product API endpoints on the Spring Boot backend
const PRODUCT_API_URL = '/api/products';

/**
 * Fetches all products, optionally supporting search/filter query parameters.
 * @param {object} query - Optional object containing search parameters (e.g., { search: 'laptop', category: 'Electronics' })
 */
const getAllProducts = (query = {}) => {
    // Convert the query object into a URL search string
    const params = new URLSearchParams(query).toString();

    // Construct the final URL
    const url = params ? `${PRODUCT_API_URL}?${params}` : PRODUCT_API_URL;

    return axios.get(url);
};

/**
 * Fetches a single product by its ID.
 * @param {number} id - The ID of the product to retrieve.
 */
const getProductById = (id) => {
    return axios.get(PRODUCT_API_URL + `/${id}`);
};

/**
 * Fetches a list of all unique categories available in the store.
 */
const getAllCategories = () => {
    return axios.get(PRODUCT_API_URL + '/categories');
};

/**
 * Fetches a list of products related to the specified product ID.
 * Assumes backend endpoint: GET /api/products/{id}/related
 * @param {number} id - The ID of the primary product.
 */
const getRelatedProducts = (id) => {
    return axios.get(PRODUCT_API_URL + `/${id}/related`);
};

/**
 * Creates a new product. Requires admin authorization.
 */
const createProduct = (productData) => {
    return axios.post(PRODUCT_API_URL, productData);
};

/**
 * Updates an existing product.
 */
const updateProduct = (id, productData) => {
    return axios.put(PRODUCT_API_URL + `/${id}`, productData);
};

/**
 * Deletes a product by ID.
 */
const deleteProduct = (id) => {
    return axios.delete(PRODUCT_API_URL + `/${id}`);
};


const ProductService = {
    getAllProducts,
    getProductById,
    getAllCategories,
    getRelatedProducts, // <-- Included for completeness
    createProduct,
    updateProduct,
    deleteProduct,
};

export default ProductService;

File: backend/src/main/java/com/ecommerce/backend/product/service/ProductService.java
package com.ecommerce.backend.product.service;

import com.ecommerce.backend.product.model.Product;
import com.ecommerce.backend.exception.ResourceNotFoundException;

import java.util.List;

/**
 * Interface for the Product Service, defining all business logic operations related to products.
 */
public interface ProductService {

    /**
     * Retrieves all products, optionally filtered by category ID.
     * @param categoryId The ID of the category to filter by (or null for all products).
     * @return A list of Product entities.
     */
    List<Product> findAllProducts(Long categoryId);

    /**
     * Retrieves a single product by its ID.
     * @param id The ID of the product.
     * @return The Product entity.
     * @throws ResourceNotFoundException if the product is not found.
     */
    Product findProductById(Long id) throws ResourceNotFoundException;

    /**
     * Creates a new product.
     * @param product The Product entity to create.
     * @return The saved Product entity, including the generated ID.
     */
    Product createProduct(Product product);

    /**
     * Updates an existing product.
     * @param id The ID of the product to update.
     * @param productDetails The Product entity containing updated details.
     * @return The updated Product entity.
     * @throws ResourceNotFoundException if the product is not found.
     */
    Product updateProduct(Long id, Product productDetails) throws ResourceNotFoundException;

    /**
     * Deletes a product by its ID.
     * @param id The ID of the product to delete.
     * @throws ResourceNotFoundException if the product is not found.
     */
    void deleteProduct(Long id) throws ResourceNotFoundException;
}
